### PR TITLE
packaging: Automate the package release process.

### PR DESCRIPTION
**Pull request by eeshangarg**
@timabbott: This is a preliminary version of how I hope to automate the release of all of our packages. Just so you know, this still needs a *lot* of work. This is just an initial implementation I wrote after school today in a few hours. I spent some of my time looking for ways to invoke `setup.py` scripts internally. Here are some things I hope to add to this:

* Have a universal `version.py` that all `setup.py` files can import from and only have to specify a new version in one place before a new release. Add a function to this script to modify that `version.py` before making a release. Ideally, after this all I should need to do is to push a commit to `python-zulip-api`.
* Have a `--cleanup` option that allows me to delete all old `build`, `dist` and the like.
* Add a `RELEASE_MANIFEST.in` specifying which files to include/exclude in a PyPA release. The script should rename said file to `MANIFEST.in` and *then* create distributions for the `zulip_bots` package.
* Have a `--beta` option that allows us to upload all packages to `<package_name>-beta` on PyPA for testing. As you mentioned, if I manage to automate this efficiently, we shouldn't need to be doing separate beta releases, but it would be a nice little add-on to have.
* Have a handy little script in the main repo called `update-zulip-package-deps` that automatically increases the `PROVISION_NUMBER`, and updates the `requirements.txt` to install off a newer release tag or commit.

I want the whole release process to boil down to:
1. Run `python-zulip-api/tools/provision`
2. Run `tools/release_packages.py`, push a commit updating `version.py` to `python-zulip-api`
3. Run `zulip/tools/update_zulip_package_deps`, submit a PR.
4. Make a post in `#integrations` notifying everyone of the new release!

This should reduce the time it takes to make a release to less than 5 minutes (I'm hoping!) :)

Looking forward to hearing your feedback! Thanks!

\-Eeshan

### CODE REVIEW

File: zulip_bots/setup.py [hunk 1]
```diff
@@ -3,14 +3,27 @@
...
 import generate_manifest
 
+def version():
+    # type: () -> str
+    version_py = os.path.join(os.path.dirname(__file__), "zulip_bots", "__init__.py")
+    with open(version_py) as in_handle:
+        version_line = next(itertools.dropwhile(lambda x: not x.startswith("__version__"),
+                                                in_handle))
+    version = version_line.split('=')[-1].strip().replace('"', '')
+    return version
```
Comment (timabbott) [inline zulip_bots/setup.py]: why aren't you just importing the file?

Comment (eeshangarg) [reply zulip_bots/setup.py]: Well, no reason, except that in `zulip/__init__.py`, we do the exact same thing to find the version, so I assumed that there might be a reason behind why we had opted to go that route so I wanted to do it the same way in the other ones. I'm guessing there is no good reason to stick with this verbose approach, so I'll just import it. :) Thanks!

Comment (timabbott) [reply zulip_bots/setup.py]: I think that's because in that package, we need the dependencies installed to import `__init__.py`, aka the main module, and that is needed to import `zulip.version`.  I suspect a more elegant fix would be to use a symlink to it in the `setup.py` directory and having `setup.py` import it from that path.

Comment (eeshangarg) [reply zulip_bots/setup.py]: Awesome, I'll do that! Thanks! :)

File: tools/release_packages [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+                        default=None,
+                        help=('Commit and push a commit changing package versions'
+                              ' (can be either "origin" or "upstream"). If "origin'
+                              ' is specified, a new branch named release-<version> is'
+                              ' checked out before committing and pushing. If'
+                              ' "upstream" is supplied, then master is checked out'
+                              ' before committing and pushing. The process is the'
+                              ' same for changes made to both repos.'))
+
+    parser.add_argument('--update-requirements',
```
Comment (derAnfaenger) [inline tools/release_packages]: Since it's really only the zulip repo we want to update, I'd rename this to `update-zulip-main-repo` or something.

File: tools/release_packages [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+    _edit_reqs_file(
+        prod_lock,
+        url_zulip_bots.format(tag=hash_or_tag, name='zulip_bots', version=version),
+        url_zulip.format(tag=hash_or_tag, name='zulip', version=version),
+    )
+    _edit_reqs_file(
+        dev_lock,
+        url_zulip_bots.format(tag=hash_or_tag, name='zulip_bots', version=version),
+        url_zulip.format(tag=hash_or_tag, name='zulip', version=version),
+    )
```
Comment (derAnfaenger) [inline tools/release_packages]: Do a `map` for both `_edi_reqs_file` calls?

File: tools/release_packages [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+                version_num = float(line.split('=')[-1].strip().replace("'", ''))
+                version_num = version_num + 0.01
+                new_file.write("PROVISION_VERSION = '{}'\n".format(version_num))
+            else:
+                new_file.write(line)
+
+    os.remove(version_py)
+    shutil.move(temp_abs_path, version_py)
+
+    message = 'Incremented PROVISION_VERSION in the main repo.'
```
Comment (derAnfaenger) [inline tools/release_packages]: I'd make this method throw an exception if no `PROVISION_VERSION` line was found.

File: tools/release_packages [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+
+def commit_and_push_version_changes(version, init_files, upstream_or_origin):
+    message = 'Committing version number changes...{}'.format(version)
+    print(crayons.yellow(message, bold=True))
+
+    if upstream_or_origin == 'origin':
+        branch = 'release-{}'.format(version)
+        git.checkout('-b', branch)
+    else:
+        branch = 'master'
```
Comment (derAnfaenger) [inline tools/release_packages]: Hm, I'm a little wary of this possibility, it seems to be potentially hazardous. Even though it makes the release process even shorter, would it be safer to make every release through PRs (one can merge them oneself, if needed). I see the danger in people with writing permission trying out this script.

File: tools/release_packages [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+    with os.fdopen(fh, 'w') as new_file, open(init_file) as old_file:
+        for line in old_file:
+            if line.startswith('__version__'):
+                prev = line.split('=')[-1].strip().replace('"', '')
+                new_file.write(line.replace(prev, new_version))
+            else:
+                new_file.write(line)
+
+    os.remove(init_file)
+    shutil.move(temp_abs_path, init_file)
```
Comment (derAnfaenger) [inline tools/release_packages]: Can we reuse this substitution code snippet for `set_git_repo_var`?

File: zulip_bots/generate_manifest.py [hunk 7]
```diff
@@ -42,17 +44,18 @@ def get_assets():
...
+    with open(MANIFEST_PATH, 'w') as fp:
         template = 'include {line}\n'
         fixtures = map(lambda line: template.format(line=line),
                        get_test_fixtures())
         logos = map(lambda line: template.format(line=line),
                     get_logos())
         docs = map(lambda line: template.format(line=line),
                    get_docs())
+        assets = map(lambda line: template.format(line=line),
+                     get_assets())
```
[the remark this answers is not in the document]
Comment (eeshangarg) [reply zulip_bots/generate_manifest.py]: Hmm, not sure, could be! But I wanted to avoid the functional changes that separated two MANIFEST files from the changes that are independent of that separation. Does that make any sense?

File: zulip_bots/generate_manifest.py [hunk 8]
```diff
@@ -63,5 +66,51 @@ def main():
...
+        docs = map(lambda line: template.format(line=line),
+                   get_docs())
+        assets = map(lambda line: template.format(line=line),
+                     get_assets())
+
+        fp.writelines(docs)
+        fp.write('\n')
+
+        fp.writelines(assets)
+        fp.write('\n')
```
[the remark this answers is not in the document]
Comment (eeshangarg) [reply zulip_bots/generate_manifest.py]: Agreed, I'll do that! Thanks! :)

File: tools/provision [hunk 9]
```diff
@@ -56,6 +62,10 @@ the Python version this command is executed with."""
             raise OSError("The command `pip install -r {}` failed. Dependencies not installed!"
                           .format(os.path.join(base_dir, requirements_filename)))
 
+    # Generate MANIFEST.in for zulip_bots
+    print("Generating MANIFEST.in for zulip_bots...")
+    generate_manifest.generate_dev_manifest()
+
```
[the remark this answers is not in the document]
Comment (eeshangarg) [reply tools/provision]: Actually, this is now redundant and can be removed, see [this](https://github.com/zulip/python-zulip-api/pull/124/files#diff-ef2d048f11abec93e29feab9f1e1a816R20). Thanks for pointing this out! :)

File: zulip_bots/setup.py [hunk 10]
```diff
@@ -3,14 +3,24 @@
...
 from __future__ import print_function
 
+import os
 import sys
 
 import generate_manifest
 
+version_py = os.path.join(os.path.dirname(__file__), "zulip_bots", "__init__.py")
+version_symlink = os.path.join(os.path.dirname(__file__), 'version.py')
+if os.path.islink(version_symlink):
```
[the remark this answers is not in the document]
Comment (eeshangarg) [reply zulip_bots/setup.py]: Because if another link was created during a previous run of setup.py, this code will fail very loudly and the setup won't succeed, which disrupts the flow of the release script in a there's-no-going-back-from-this fashion. In that case we should remove the pre-existing link and start anew! :)

File: zulip_bots/setup.py [hunk 11]
```diff
@@ -36,7 +46,7 @@
 
 setuptools_info = dict(
     install_requires=[
-        'zulip>=0.3.4',
+        'zulip>={}'.format(ZULIP_BOTS_VERSION),
```
[the remark this answers is not in the document]
Comment (eeshangarg) [reply zulip_bots/setup.py]: From my experience as a maintainer so far, I have never had to make separate releases for these packages with separate version numbers. For now, I think we should stick with one version for all packages, since they will be released simultaneously (ideally). Actually, I wanted to have a universal `version.py` in the repo root directory, but the way packaging works, that wouldn't have worked!

I don't think we should over-complicate this by writing a script that tries to get the version number from an adjoining package, that wouldn't work anyway because each package is its own distinct self (so to speak) and shouldn't depend on anything outside of it *before* it has been installed.

I do agree with you on one thing that they are technically independent, so let's just skip the version and say `zulip` instead of `zulip>=<version>`. The former invocation would install the latest package automatically, so we don't need to explicitly specify it here. :) Thanks for pointing this out, Robert!

File: zulip_bots/setup.py [hunk 12]
```diff
@@ -66,7 +76,7 @@ def check_dependency_manually(module_name, version=None):
                 print("{name} is not installed.".format(name=module_name), file=sys.stderr)
             sys.exit(1)
 
-    check_dependency_manually('zulip', '0.3.4')
+    check_dependency_manually('zulip', ZULIP_BOTS_VERSION)
```
[the remark this answers is not in the document]
Comment (eeshangarg) [reply zulip_bots/setup.py]: I replied to your feedback about this above. :)

File: zulip_botserver/setup.py [hunk 13]
```diff
@@ -30,8 +40,8 @@
 
 setuptools_info = dict(
     install_requires=[
-        'zulip>=0.3.4',
-        'zulip_bots>=0.3.4',
+        'zulip>={}'.format(ZULIP_BOTSERVER_VERSION),
+        'zulip_bots>={}'.format(ZULIP_BOTSERVER_VERSION),
```
[the remark this answers is not in the document]
Comment (eeshangarg) [reply zulip_botserver/setup.py]: I replied to your feedback about this above. :)

File: requirements.txt [hunk 14]
```diff
@@ -1,3 +1,6 @@
+crayons
```
[the remark this answers is not in the document]
Comment (eeshangarg) [reply requirements.txt]: Are you sure about this? `crayons` is built on top of `colorama`, see [here](https://github.com/kennethreitz/crayons/blob/master/setup.py#L28). So shouldn't it work on Windows?

File: zulip_bots/setup.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (derAnfaenger) [reply zulip_bots/setup.py]: My reasoning was that eventually, some package like `zulip_botserver` might be very stable and doesn't need to be updated, whereas frequent updates occur for other packages. In such a case, I think it'd be more convenient for users of the packages to be able to not worry about three different package version when they are in fact all the same. Maybe we can discuss this at today's meeting :)

File: tools/release_packages (hunk 6 above)
Comment (eeshangarg) [reply tools/release_packages]: Hmm.. we could, yes! I'll get that done! Thanks, Robert! :)

File: requirements.txt (hunk 14 above)
[the remark this answers is not in the document]
Comment (derAnfaenger) [reply requirements.txt]: Oh, interesting. Yup, then it should work on Windows.

File: tools/release_packages (hunk 5 above)
Comment (eeshangarg) [reply tools/release_packages]: Well, I think it doesn't matter for the following reasons:
* Anyone without PyPA credentials shouldn't be using this script.
* For changes that can be pushed to the main repo, I trust anyone who has commit access to that repo to exercise their judgement and use `--push origin` and *not* `--push upstream`. If Tim or Steve or Greg want to push to master, they should be free to do so.
* For people who have commit access to this repo, I also trust them to not make a release without consulting me or senior developers like Tim (unless I am mistaken, I think Tim and I are the only ones with PyPA credentials for making a release).
* Even if someone were to push a commit to this repo changing the version numbers, that wouldn't necessarily break anything anywhere. The people installing the package from PyPA would get what they are supposed to get. And since the main repo installs off of a tag or commit, it would get what it wants regardless.

To sum up, I think this is okay to do. If you'd like to discuss this further, let me know! :)

File: tools/release_packages (hunk 3 above)
Comment (eeshangarg) [reply tools/release_packages]: Agreed, much more concise! Thanks, Robert! :)

File: tools/release_packages (hunk 5 above)
Comment (derAnfaenger) [reply tools/release_packages]: Okay, agreed.

File: tools/release_packages (hunk 4 above)
Comment (eeshangarg) [reply tools/release_packages]: I don't think there is a scenario where `PROVISION_VERSION` won't exist. Shouldn't `zulip/version.py` always have it? What do you think? :)

File: tools/release_packages [hunk 15]
```diff
@@ -0,0 +1,372 @@
...
+        zulip_git.add(common, prod_lock, dev_lock, version_py)
+
+        commit_msg = 'requirements: Upgrade to version {} of the Zulip API packages.'.format(version)
+        print(crayons.yellow('Commit message: {}'.format(commit_msg), bold=True))
+        zulip_git.commit('-m', commit_msg)
+
+        zulip_git.push(upstream_or_origin, branch)
+
+def set_git_repo_var(fp, to=False):
+    fh, temp_abs_path = tempfile.mkstemp()
```
[the remark this answers is not in the document]
Comment (eeshangarg) [reply tools/release_packages]: Sure, I can do that. :)

File: tools/release_packages [hunk 16]
```diff
@@ -0,0 +1,372 @@
...
+4. --update-requirements: Update the requirements/ in the main
+   zulip repo to install off of the newest version of the packages.
+   Also increments PROVISION_VERSION in the main repo.
+5. --push origin: Commit the changes produced in Step 1 and 4,
+   checkout a new branch named release-<version>, generate a commit
+   message, commit the changes and push the branch to origin.
+"""
+    parser = argparse.ArgumentParser(usage=usage)
+
+    parser.add_argument('release_version',
```
[the remark this answers is not in the document]
Comment (eeshangarg) [reply tools/release_packages]: I replied to your concerns about the version in a comment above. :) Ideally, yes! I suggest we keep it simple till we get to a point where the packages start to diverge in ways that prompt us to maintain each one separately. However, if you disagree, I'd love to hear more about it! :)

File: tools/release_packages (hunk 2 above)
Comment (eeshangarg) [reply tools/release_packages]: Sure, I'll do that. :)

File: tools/release_packages [hunk 17]
```diff
@@ -0,0 +1,372 @@
...
+                              ' (can be either "origin" or "upstream"). If "origin'
+                              ' is specified, a new branch named release-<version> is'
+                              ' checked out before committing and pushing. If'
+                              ' "upstream" is supplied, then master is checked out'
+                              ' before committing and pushing. The process is the'
+                              ' same for changes made to both repos.'))
+
+    parser.add_argument('--update-requirements',
+                        metavar='PATH_TO_ZULIP_DIR',
+                        default=None,
```
[the remark this answers is not in the document]
Comment (eeshangarg) [reply tools/release_packages]: Actually, I am not sure if it defaults to an empty string for string parameters or to None. I'll look into it. Thanks for pointing this out! :)

File: tools/release_packages (hunk 4 above)
Comment (derAnfaenger) [reply tools/release_packages]: I don't think it'll happen either, it's just that right now, for any reason we might or might not thinkg of, a missing `PROVISION_VERSION` would fail silently, and an exception is easy and quick to add here.

File: tools/release_packages [hunk 18]
```diff
@@ -0,0 +1,372 @@
...
+                )
+            generate_sdist(setup_file, package_name)
+            generate_bdist_wheel_universal(setup_file, package_name)
+
+    if options.release:
+        dist_dirs = glob.glob(os.path.join(REPO_DIR, '*', 'dist'))
+        twine_upload(dist_dirs)
+
+    if options.update_requirements is not None:
+        if options.hash:
```
[the remark this answers is not in the document]
Comment (eeshangarg) [reply tools/release_packages]: I guess this relates to your comment above, I'll definitely take a look at this. :)

File: zulip_bots/zulip_bots/__init__.py [hunk 19]
```diff
@@ -1 +1,2 @@
 __version__ = "0.3.4"
+GIT_REPO = True
```
[the remark this answers is not in the document]
Comment (eeshangarg) [reply zulip_bots/zulip_bots/__init__.py]: Yep, that's much better! Consider it done! :)

File: .codecov.yml [hunk 20]
```diff
@@ -8,3 +8,6 @@ coverage:
         threshold: 0.85
         base: auto
     patch: off
+
+ignore:
+  - "[a-z]+/setup.py"  # Ignore all setup.py files
```
[the remark this answers is not in the document]
Comment (eeshangarg) [reply .codecov.yml]: I actually had a discussion about this with Tim before going ahead and ignoring setup.py files explicitly. Here are the reasons:
* `setup.py` files will contain a lot of boilerplate that is hard to test and usually is never tested. Also, with our automation code, this increases the lines of boilerplate needed to install the packages in different ways.
* We don't necessarily need to optimize *every* file for test coverage, certainly not `setup.py` files or scripts like `release_packages`.
* The build was failing because of codecov, so it didn't seem right to merge something that was explicitly failing. So this seemed like a good way to say that its intended to fail.

What do you think? :)

File: tools/release_packages (hunk 4 above)
Comment (eeshangarg) [reply tools/release_packages]: Cool, I agree, I'll do that then! :)

File: zulip_bots/setup.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (eeshangarg) [reply zulip_bots/setup.py]: Okay, I emphasized this in a comment further down. I agree with you on that. I just don't think that separation is something that needs to be done as a part of this PR. That could probably be done later on when we realize that the packages are diverging in terms of stability and versioning scheme. :)

File: .codecov.yml (hunk 20 above)
[the remark this answers is not in the document]
Comment (derAnfaenger) [reply .codecov.yml]: I don't understand this part:
> Also, with our automation code, this increases the lines of boilerplate needed to install the packages in different ways.

Apart from that, I agree with the first two points.

For the third point, I think it's fine if codecov fails for the build, since it only tests coverage and we know that the reason is negligible in this case. Future builds won't fail, since their coverage change will be calculated relative to this commit. Furthermore we aren't near a coverage percentage where it would make sense to exclude files for the sake of keeping the number high. 

All in all, I don't really see a reason to exclude it. I also don't see any reason not to exclude it, other than three additional lines. Thus, I'd vote for keeping codecov as it is, but it's not a strong opinion.

[the remark this answers is not in the document]
Comment (eeshangarg) [reply .codecov.yml]: Cool, I'm convinced, we won't exclude it then! :)

What I meant by:
``` quote
Also, with our automation code, this increases the lines of boilerplate needed to install the packages in different ways.
```
is that with the release script, I had to add a bunch of boilerplate to `setup.py` files that triggered the codecov warnings, because that code wasn't tested.

**Comment by eeshangarg** [comment]
@timabbott: @derAnfaenger and I decided not to exclude the `setup.py` files in codecov, so codecov failing was an intended consequence in this case! :) Apart from that, I think this is ready to be merged.

File: zulip_botserver/version.py [hunk 21]
```diff
@@ -0,0 +1 @@
+/home/eeshangarg/python-zulip-api/zulip_botserver/zulip_botserver/__init__.py
```
Comment (timabbott) [inline zulip_botserver/version.py]: You wanted to make a relative link, not an absolute link, with e.g.

```
ln -nsf ../zulip_botserver/zulip_botserver/__init__.py version.py 
```

File: zulip_botserver/setup.py [hunk 22]
```diff
@@ -3,12 +3,22 @@
...
 
+import os
 import sys
 
+version_py = os.path.join(os.path.dirname(__file__), "zulip_botserver", "__init__.py")
+version_symlink = os.path.join(os.path.dirname(__file__), 'version.py')
+if os.path.islink(version_symlink):
+    os.remove(version_symlink)
+os.symlink(version_py, version_symlink)
+import version
```
Comment (timabbott) [inline zulip_botserver/setup.py]: Why is this created dynamically?  You should be able to just use a static relative path symlink.

**Comment by timabbott** [comment]
Posted a few comments.  My main concern, other than that this is a lot of code, is that I'm not sure why you're making the symlinks where you are.

**Comment by timabbott** [comment]
Merged, thanks @eeshangarg!  Please be sure to carefully test the contents of the various packages before doing a release, especially the `zulip` one (since that one is much more widely used), but I think we should probably do a release soon.
